### PR TITLE
Assignment 6 : Testing For Assignment 3 (Inventory Manager)

### PR DESCRIPTION
**Pull request by DhivakarG13**
### Justification

- Did X_Unit testing for Validation methods 
- Search methods are tested
- Methods setting the value of property tested

### To_Do
- [x] Test with More Edge cases.
- [x] Add tests to more methods if missed.

### CODE REVIEW

File: Assignment_3_xUnitTests/UnitTest1.cs
Comment (Kaiser-Asan) [inline Assignment_3_xUnitTests/UnitTest1.cs:1]: Remove unwanted files

File: Assignment_3_xUnitTests/InventoryOperations/SearchOperationsTests.cs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        [InlineData(2005, 1)]
+        [InlineData(2006, 2)]
+        [InlineData(2007, 3)]
+        [InlineData(2008, 4)]
+        [InlineData(2009, 5)]
+        public void GivenProductsAndProductId_WhenSearchByProductId_ThenReturnsListOfMatchingProducts(int productIDToSearch, int matchingIndex)
+        {
+            List<Product> testProducts = GenerateProducts(6);
+            List<Product> expectedProducts = new List<Product>();
+            expectedProducts.AddRange(new Product[] { testProducts[matchingIndex] }); //new Product("Jester", 28, 200, 2008, DateOnly.MinValue),
```
Comment (Kaiser-Asan) [inline Assignment_3_xUnitTests/InventoryOperations/SearchOperationsTests.cs:16]: no need for List<Product> actualProducts = new List<Product>()

File: Assignment_3_xUnitTests/InventoryOperations/SearchOperationsTests.cs [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+            actualProducts = InventoryOperations.SearchByProductId(testProducts, productIDToSearch);
+
+            Assert.Empty(actualProducts);
+        }
+
+        [Theory]
+        [InlineData(200, 201)]
+        public void GivenProductsAndMinAndMaxOfRange_WhenSearchByProductPrizeRange_ThenReturnsListOfProductsInTheRange(int minValue = 200, int maxValue = 400)
+        {
```
Comment (Kaiser-Asan) [inline Assignment_3_xUnitTests/InventoryOperations/SearchOperationsTests.cs:44]: Fix spelling ProductPrizeRange -> ProductPriceRange

File: Assignment_3_xUnitTests/InventoryOperations/SearchOperationsTests.cs [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+            List<Product> expectedProducts = new List<Product>();
+            expectedProducts.AddRange(new Product[] { testProducts[1],   
+                                                      testProducts[10]   
+                                                    });
+
+            List<Product> actualProducts = new List<Product>();
+            actualProducts = InventoryOperations.SearchByProductName(testProducts, productNameToSearch);
+
+            Assert.Equal(actualProducts, expectedProducts);
```
Comment (Kaiser-Asan) [inline Assignment_3_xUnitTests/InventoryOperations/SearchOperationsTests.cs:83]: Fix order. always (expected,actual)

File: Assignment_3_xUnitTests/InventoryOperations/SearchOperationsTests.cs
Comment (Kaiser-Asan) [inline Assignment_3_xUnitTests/InventoryOperations/SearchOperationsTests.cs:1]: Test for product name partial match?

Comment (Kaiser-Asan) [reply Assignment_3_xUnitTests/InventoryOperations/SearchOperationsTests.cs:1]: Empty product name test

Comment (Kaiser-Asan) [reply Assignment_3_xUnitTests/InventoryOperations/SearchOperationsTests.cs:1]: Product search tests when list of products empty

File: Assignment_3_xUnitTests/InventoryOperations/SearchOperationsTests.cs [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+            List<Product> actualProducts = InventoryOperations.SearchByProductPrizeRange(testProducts, minValue, maxValue);
+
+            Assert.Equal(actualProducts, expectedProducts);
+        }
+
+        [Theory]
+        [InlineData(2000, 4000)]
+        [InlineData(3000, 4000)]
+        [InlineData(9000, 10000)]
+        public void GivenProductsAndMinAndMaxOfRange_WhenSearchByProductPrizeRange_ThenReturnsEmptyListIfNoProductsInTheRange(int minValue = 200, int maxValue = 400)
```
Comment (Kaiser-Asan) [inline Assignment_3_xUnitTests/InventoryOperations/SearchOperationsTests.cs:60]: why assign values to min and max if passing inline?

File: Assignment_3_xUnitTests/InventoryOperations/SearchOperationsTests.cs
Comment (Kaiser-Asan) [reply Assignment_3_xUnitTests/InventoryOperations/SearchOperationsTests.cs:1]: Test when multiple products have same name.

Comment (Kaiser-Asan) [reply Assignment_3_xUnitTests/InventoryOperations/SearchOperationsTests.cs:1]: Product name case sensitive match test

File: Assignment_3_xUnitTests/Utilities/ValidationServiceTests.cs [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+            Assert.False(expectedResult);
+        }
+
+        [Theory]
+        [InlineData("0")]
+        [InlineData("700")]
+        [InlineData("823000")]
+        [InlineData("009")]
+        [InlineData("010")]
+        public void GivenString_WhenValidatePrice_ThenReturnsTrueIfValidPrice(string numericalValue)
```
Comment (Kaiser-Asan) [inline Assignment_3_xUnitTests/Utilities/ValidationServiceTests.cs:102]: Validate double values for price?

File: Assignment_3_xUnitTests/Utilities/ValidationServiceTests.cs [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+            Assert.True(ExpectedValue);
+        }
+
+        [Theory]
+        [InlineData("")]
+        [InlineData(null)]
+        [InlineData("Hi")]
+        [InlineData("0")]
+        [InlineData(" 9 ")]
+        public void GivenStringValue_WhenValidateStringValue_ThenReturnsFalseIfStringLengthLesserThan3(string stringValue)
```
Comment (Kaiser-Asan) [inline Assignment_3_xUnitTests/Utilities/ValidationServiceTests.cs:193]: Validate names with mixed characters  , String with symbols

File: Assignment_3_xUnitTests/Utilities/ValidationServiceTests.cs [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+        }
+
+        [Theory]
+        [InlineData("0")]
+        [InlineData("7")]
+        [InlineData("8")]
+        [InlineData("9")]
+        [InlineData("10")]
+        [InlineData("11")]
+        public void GivenChoiceAndListOfChoices_WhenValidateChoice_ThenReturnsFalseIfOutOfRangeChoice(string userChoice)
```
Comment (Kaiser-Asan) [inline Assignment_3_xUnitTests/Utilities/ValidationServiceTests.cs:31]: Add cases for leading, trailing space ("  1" or "1  ")

**Review by Kaiser-Asan — CHANGES_REQUESTED** [review]

File: Assignment_3_xUnitTests/Utilities/ValidationServiceTests.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (DhivakarG13) [reply Assignment_3_xUnitTests/Utilities/ValidationServiceTests.cs]: changes made in name of methods

File: Assignment_3_xUnitTests/Utilities/ValidationServiceTests.cs [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+            Assert.False(actualResult);
+        }
+
+        [Theory]
+        [InlineData("0")]
+        [InlineData("700")]
+        [InlineData("823000")]
+        [InlineData("009")]
+        [InlineData("010")]
+        public void GivenString_WhenValidateNumericalInputs_ThenReturnsTrueIfPositiveNumber(string numericalValue)
```
[the remark this answers is not in the document]
Comment (DhivakarG13) [reply Assignment_3_xUnitTests/Utilities/ValidationServiceTests.cs]: GivenString -> string to validate

File: Assignment_3_xUnitTests/Utilities/IdGenerateServiceTests.cs [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Inventory.Helpers;
+
+namespace Assignment_3_xUnitTests.Utilities
+{
+    public class IdGenerateServiceTests
+    {
+        [Theory]
```
[the remark this answers is not in the document]
Comment (DhivakarG13) [reply Assignment_3_xUnitTests/Utilities/IdGenerateServiceTests.cs:12]: Used Linq in assert

File: Assignment_3_xUnitTests/InventoryOperations/SearchOperationsTests.cs (hunk 4 above)
Comment (DhivakarG13) [reply Assignment_3_xUnitTests/InventoryOperations/SearchOperationsTests.cs]: optional parameters removed

File: Assignment_3_xUnitTests/Utilities/ValidationServiceTests.cs (hunk 5 above)
Comment (DhivakarG13) [reply Assignment_3_xUnitTests/Utilities/ValidationServiceTests.cs]: Price is integer type

File: Assignment_3_xUnitTests/Utilities/ValidationServiceTests.cs (hunk 6 above)
Comment (DhivakarG13) [reply Assignment_3_xUnitTests/Utilities/ValidationServiceTests.cs]: String accepts special character, Added tests for that too

File: Assignment_3_xUnitTests/Utilities/ValidationServiceTests.cs (hunk 7 above)
Comment (DhivakarG13) [reply Assignment_3_xUnitTests/Utilities/ValidationServiceTests.cs]: Added tests for trailing spaces

File: Assignment_3_xUnitTests/UnitTest1.cs
Comment (DhivakarG13) [reply Assignment_3_xUnitTests/UnitTest1.cs:1]: Unwanted files removed

File: Assignment_3_xUnitTests/InventoryOperations/SearchOperationsTests.cs (hunk 3 above)
Comment (DhivakarG13) [reply Assignment_3_xUnitTests/InventoryOperations/SearchOperationsTests.cs:83]: order rectified

File: Assignment_3_xUnitTests/Utilities/ValidationServiceTests.cs [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+
+            Assert.False(expectedResult);
+        }
+
+        [Theory]
+        [InlineData("-700")]
+        [InlineData("-823000")]
+        [InlineData("-9")]
+        [InlineData("-10")]
+        public void GivenString_WhenValidatePrice_ThenReturnsFalseIfNegativeNumber(string numericalValue)
```
[the remark this answers is not in the document]
Comment (DhivakarG13) [reply Assignment_3_xUnitTests/Utilities/ValidationServiceTests.cs]: Name changed

File: Assignment_3_xUnitTests/Utilities/ValidationServiceTests.cs [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+
+        [Theory]
+        [InlineData("")]
+        [InlineData(null)]
+        [InlineData("Ab0")]
+        [InlineData("1*!")]
+        [InlineData("Hello")]
+        [InlineData("--1")]
+        [InlineData("++100")]
+        public void GivenString_WhenValidateNumericalInputs_ThenReturnsFalseIfParseFail(string numericalValue)
```
[the remark this answers is not in the document]
Comment (DhivakarG13) [reply Assignment_3_xUnitTests/Utilities/ValidationServiceTests.cs]: Name Changed

File: Assignment_3_xUnitTests/InventoryOperations/SearchOperationsTests.cs
Comment (DhivakarG13) [reply Assignment_3_xUnitTests/InventoryOperations/SearchOperationsTests.cs:1]: Added tests for all requirements

**Review by Kaiser-Asan — CHANGES_REQUESTED** [review]
Why is assignment 2 and 3 added in this PR? Were there any changes make in assignment 3 to help with XUnit tests. 

Did you try to rebase or merge from main, If you should point this PR to Main

**Review by Kaiser-Asan — CHANGES_REQUESTED** [review]

**Review by Kaiser-Asan — APPROVED** [review]

File: Assignment_3_xUnitTests/Assignment_3_xUnitTests.csproj
[the remark this answers is not in the document]
Comment (DhivakarG13) [reply Assignment_3_xUnitTests/Assignment_3_xUnitTests.csproj:1]: Added readme file for the app
